cirq/circuits/circuit.py

```diff
@@ -1387,6 +1389,48 @@ def tetris_concat(
 
         return cirq.Circuit(buffer[offset : offset + n_acc])
 
+    def get_qubit_factors(self) -> List[Set['cirq.Qid']]:
+        """
+        Divide circuit's qubits into qubit factors such that there are
+        no entangling gates between the qubits belonging to different factors.
+        If this is not possible, a sequence with a single factor (the whole set of
+        circuit's qubits) will be returned.
+        """
+        uf = networkx.utils.UnionFind(self.all_qubits())
+        for op in self.all_operations():
+            if len(op.qubits) > 1:
+                uf.union(*op.qubits)
+        return sorted(uf.to_sets(), key=min)
+
+    def factorize(self: CIRCUIT_TYPE) -> Iterable[CIRCUIT_TYPE]:
+        """Factorize circuit into a sequence of circuits, if it's possible (i.e. if
+        the circuit qubits can be divided into two or more groups of qubits
+        such that there are no entangling gates between them).
+        If this is not possible, will return the set consisting of the single
+        circuit (this one).
+        """
+
+        qubit_factors = self.get_qubit_factors()
+        if len(qubit_factors) == 1:
+            return (self,)
+        return (
+            self._with_sliced_moments([m[qubits] for m in self.moments]) for qubits in qubit_factors
+        )
+
+    def slice_by_qubit(
+        self: CIRCUIT_TYPE, with_qubits: Iterable['cirq.Qid']
+    ) -> Iterable[CIRCUIT_TYPE]:
```

Comment:
Not sure I understand the purpose of this function. From the name, I would expect it to return a single circuit that contains gates on only the given qubits, rather than several circuits where we also factorize. In other words, something:
```python
return self._with_sliced_moments([m[with_qubits] for m in self.moments])
```
The resulting circuit could be factorized if desired, but it's not clear to me why the factorization should be done here. There's a question of what to do if the original circuit contains gates on qubits both inside and outside the specified set; in the snippet above the set would be expanded to include other qubits, but one could also imagine either raising an exception in that case, or just dropping any such gates. It might make sense to allow any of those behaviors, selected by an additional arg.

Reply:
Valid point indeed. I have changed the function to return a single circuit.

The function was introduced as a spinoff to `factorize` om mpharrigan's suggestion (return only the the circuit based on factors containing the specific qubits).